### ISSUE TITLE
Error: Cannot find module './grammar.js'? Does anyone know how to resolve this?

### ISSUE THREAD

**Issue by shadowsocksquestions** [haraka/Haraka#3054]
Hi

So I'm trying to follow the getting started guide, though encounter this error when trying to start the process

```
$ sudo haraka -c /var/www/test-mail
loaded TLD files:
  1=1495
  2=8570
  3=2448
loaded 9192 Public Suffixes
node:internal/modules/cjs/loader:942
  throw err;
  ^

Error: Cannot find module './grammar.js'
Require stack:
- /usr/local/lib/node_modules/Haraka/node_modules/address-rfc2821/index.js
- /usr/local/lib/node_modules/Haraka/connection.js
- /usr/local/lib/node_modules/Haraka/logger.js
- /usr/local/lib/node_modules/Haraka/haraka.js
- /usr/local/lib/node_modules/Haraka/bin/haraka
    at Module._resolveFilename (node:internal/modules/cjs/loader:939:15)
    at Module._load (node:internal/modules/cjs/loader:780:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/usr/local/lib/node_modules/Haraka/node_modules/address-rfc2821/index.js:6:46)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Module._load (node:internal/modules/cjs/loader:827:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/usr/local/lib/node_modules/Haraka/node_modules/address-rfc2821/index.js',
    '/usr/local/lib/node_modules/Haraka/connection.js',
    '/usr/local/lib/node_modules/Haraka/logger.js',
    '/usr/local/lib/node_modules/Haraka/haraka.js',
    '/usr/local/lib/node_modules/Haraka/bin/haraka'
  ]
}
Node.js v18.0.0
```

Here are the environment details:
Haraka | Haraka.js — Version: 2.8.28
 --- | :--- 
Node | v18.0.0
OS | Linux [my-machine-name] 5.11.0-49-generic #55-Ubuntu SMP Wed Jan 12 17:36:34 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
openssl | OpenSSL 1.1.1j  16 Feb 2021

If anyone knows how to solve this, I'd appreciate it

Thanks

**Comment by gene-hightower** [comment]
The missing file is the output of the Nearley parser generator <https://nearley.js.org/> that should run during the install phase.
Did you run "npm install" step with success?

**Comment by gene-hightower** [comment]
The grammar.js file is generated by the Nearley parser generator, <https://nearley.js.org/> which should be run in the package install step.

**Comment by msimerson** [comment]
I wonder if there's an issue with node 18/npm/nearley that causes the grammar file to not be generated during the package's installation. Try again with node 14 or 16 and very that this isn't an issue specific to node 18.

**Comment by shadowsocksquestions** [comment]
@gene-hightower 

Thanks. I figured this out too. It seems it was only that one file that was missing...everything else was there, as I state below

@msimerson 

It seems to be - with v16 I also got the same error, but with v14, it works fine. I could then re-upgrade back to v18 and simply copy the grammar.js file that was created with v14 into the new node modules global folder...and it seems everything works fine

I then spent a couple of hours trying out various configs/trying to understand the docs, though had most luck just grepping for the string "body" in the project folder and could eventually decipher out that I needed to use hook_data_post and can access the message body with connection.transaction.body.bodytext IIRC...having made sure to set parse_body to true in the hook_data function...so for the maintainers of the project, a real simple "noob demo" showing how to access all the key information - the sender, recipient, message body etc and whatever other fields are included in the email etc could be useful (the only tutorial I could find only used the hook_rcpt...and it's not at all obvious that you would need e.g. the hook_data_post to access the body of the email). Also possibly a simple tutorial showing how to send an email from one client on the localhost to a receiver on the localhost and access all the key data mentioned previously would be useful for testing purposes

If the project has funding or anything, this could be a good idea...though of course, if everyone here are volunteers (which I assume?), then obviously no one has any obligation to contribute anything. 

Thanks for the help though...and I'm excited to try this software properly over the next few days
